### PR TITLE
[SSHD-1324] Rooted file system can leak informations

### PR DESCRIPTION
**Pull request by gnodet**
- More unix like
- Symlinks improvements

### CODE REVIEW

**Comment by tomaswolf** [comment]
It would be extremely helpful to have meaningful commit messages. Just "More unix like" is _not_ helpful. If someone looks at this code a few months or years from now he or she will have no idea what this is about from the git history. There's not even any link to any issue, PR, or mailing list message. In this case I do not expect the commit message to give a full explanation, but brief descriptions should be possible.

At the very least there must be a meaningful summary line so that `git log --oneline` gives a result that correctly summarizes the scope and topic. 

Absent a JIRA or Github issue, a link back to this PR might also helpful.

Something like
```
RootedFileSystem: more Unix-like path handling

Clean up and fix path checks to be more like in Unix.
Add unit tests for Unix, OSX, and Windows via JimFs.
See also [1].

[1] https://github.com/apache/mina-sshd/pull/362
```
should be the minimum for the first commit. And likewise for the second commit.

File: sshd-common/src/test/java/org/apache/sshd/common/file/root/RootedFileSystemProviderTest.java [hunk 1]
```diff
@@ -20,60 +20,118 @@
...
+            } catch (Exception ex) {
+                failures.add(ex);
+            }
+        }
+        Assert.assertEquals(Collections.emptyList(), failures);
+    }
+
+    private Path getTargetFolderOnHostFs(Path targetFolder) {
+        return this.hostFilesystem.getSeparator().equals("\\")
+                ? this.hostFilesystem.getPath("C:", targetFolder.toString())
```
Comment (tomaswolf) [inline sshd-common/src/test/java/org/apache/sshd/common/file/root/RootedFileSystemProviderTest.java:134]: This looks suspicious. In CI, we're on drive D:.

File: sshd-common/src/main/java/org/apache/sshd/common/util/io/IoUtils.java [hunk 2]
```diff
@@ -637,4 +677,75 @@ public static List<String> readAllLines(BufferedReader reader, int lineCountHint
...
+    private static Path buildPath(Path root, FileSystem fs, List<String> namesList) {
+        if (namesList.isEmpty()) {
+            return root == null ? fs.getPath(".") : root;
+        }
+
+        Path cleanedPathToResolve = buildPath(fs, namesList);
+        return root.resolve(cleanedPathToResolve);
+    }
+
+    private static Path buildPath(FileSystem fs, List<String> namesList) {
```
[the remark this answers is not in the document]
Comment (pyckle) [reply sshd-common/src/main/java/org/apache/sshd/common/util/io/IoUtils.java:740]: A List is an ordered collection. Why should this API permit an unordered collection?

File: sshd-common/src/main/java/org/apache/sshd/common/util/io/IoUtils.java [hunk 3]
```diff
@@ -637,4 +677,75 @@ public static List<String> readAllLines(BufferedReader reader, int lineCountHint
...
+                if (!".".equals(name)) {
+                    ++numDirParts;
+                }
+                newNames.add(name);
+            }
+        }
+        return newNames;
+    }
+
+    private static Path buildPath(Path root, FileSystem fs, List<String> namesList) {
```
[the remark this answers is not in the document]
Comment (pyckle) [reply sshd-common/src/main/java/org/apache/sshd/common/util/io/IoUtils.java:731]: A List is an ordered collection. Why should this API permit an unordered collection?

[the remark this answers is not in the document]
Comment (pyckle) [reply sshd-common/src/main/java/org/apache/sshd/common/util/io/IoUtils.java:731]: I can agree with this approach in many situations, but there's no practical use case here. Having a `Set` (or any other unordered collection for that matter) for a set of path components makes absolutely no sense. The `Collection` type in the signature only will confuse a developer as the signature doesn't semantically reflect what the function accomplishes.

And if this scenario was a use case (which I would be very curious to see), calling `buildPath(..., new ArrayList(namesCollection),...)` would be an easy way to use this function.

With that said, I'm not a maintainer, and I made a PR for @gnodet which makes this change (and fixes the other issues) here: https://github.com/gnodet/mina-sshd/pull/2/files

**Comment by pyckle** [comment]
I made the suggested changes here: https://github.com/gnodet/mina-sshd/pull/2/files

**Comment by gnodet** [comment]
@tomaswolf I plan to squash the commits into a single commit

**Comment by tomaswolf** [comment]
@gnodet: seems this needs more work. I suggest we publish a 2.10.0 without this change, and do a 2.10.1 once this is ready. I'd really like to have 2.10.0 in Eclipse 2023-06, but to make that work, I'll need a release next week. Otherwise it'll be too late in the Eclipse release process.

**Comment by gnodet** [comment]
> We'd greatly appreciate it if you could give us some advice on whether the [CVE-2023-35887](https://github.com/advisories/GHSA-mjmq-gwgm-5qhm) vulnerability affects Apache MINA 2.1.X and 2.2.X?

The issue affects the Apache Mina *SSHD* project, not the Apache Mina library.

**Comment by 07070529** [comment]
Thanks a lot for your reply, and we appreciate your prompt attention to this issue.

**Pull request merged**